resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocatorBlacklistTracker.scala

```diff
@@ -0,0 +1,153 @@
+
+  private var schedulerBlacklistedNodesWithExpiry = Map.empty[String, Long]
+
+  private var numClusterNodes = (Int.MaxValue / BLACKLIST_MAX_NODE_BLACKLIST_RATIO).toInt
+
+  def setNumClusterNodes(numClusterNodes: Int): Unit = {
+    this.numClusterNodes = numClusterNodes
+  }
+
+  /**
+   * Use a different clock. This is mainly used for testing.
+   */
+  def setClock(newClock: Clock): Unit = {
+    clock = newClock
+  }
+
+  def handleResourceAllocationFailure(hostOpt: Option[String]): Unit = {
+    hostOpt match {
+      case Some(hostname) =>
+        // failures on a already blacklisted nodes are not even tracked
+        // otherwise such failures could shutdown the application
+        // as resource requests are asynchronous
+        // and a late failure response could exceed MAX_EXECUTOR_FAILURES
+        if (!schedulerBlacklistedNodesWithExpiry.contains(hostname) &&
+          !allocationBlacklistedNodesWithExpiry.contains(hostname)) {
+          failureWithinTimeIntervalTracker.registerFailureOnHost(hostname)
+          updateAllocationBlacklistedNodes(hostname)
+        }
+      case None =>
+        failureWithinTimeIntervalTracker.registerExecutorFailure()
+    }
+  }
+
+  private def updateAllocationBlacklistedNodes(hostname: String): Unit = {
+    if (IS_YARN_EXECUTOR_LAUNCH_BLACKLIST_ENABLED) {
+      val failuresOnHost = failureWithinTimeIntervalTracker.getNumExecutorFailuresOnHost(hostname)
+      if (failuresOnHost > BLACKLIST_MAX_FAILED_EXEC_PER_NODE) {
+        logInfo(s"blacklisting $hostname as YARN allocation failed $failuresOnHost times")
+        allocationBlacklistedNodesWithExpiry.put(
+          hostname,
+          clock.getTimeMillis() + BLACKLIST_TIMEOUT_MILLIS)
+        refreshBlacklistedNodes()
+      }
+    }
+  }
+
+  def setSchedulerBlacklistedNodes(schedulerBlacklistedNodesWithExpiry: Map[String, Long]): Unit = {
+    this.schedulerBlacklistedNodesWithExpiry = schedulerBlacklistedNodesWithExpiry
+    refreshBlacklistedNodes()
+  }
+
+  private def refreshBlacklistedNodes(): Unit = {
+    removeExpiredYarnBlacklistedNodes()
+    val limit = (numClusterNodes * BLACKLIST_MAX_NODE_BLACKLIST_RATIO).toInt
+    val nodesToBlacklist =
+      if (schedulerBlacklistedNodesWithExpiry.size +
+          allocationBlacklistedNodesWithExpiry.size > limit) {
+        mostRelevantSubsetOfBlacklistedNodes(limit)
+      } else {
+        schedulerBlacklistedNodesWithExpiry.keySet ++ allocationBlacklistedNodesWithExpiry.keySet
+      }
+
+    synchronizeBlacklistedNodeWithYarn(nodesToBlacklist)
+  }
+
+  private def mostRelevantSubsetOfBlacklistedNodes(limit: Int) = {
+    val relevant =
+      (schedulerBlacklistedNodesWithExpiry ++ allocationBlacklistedNodesWithExpiry).toSeq
+      .sortBy(_._2)(Ordering[Long].reverse)
+      .take(limit)
+      .map(_._1)
+      .toSet
+    logInfo(s"blacklist size limit ($limit) is reached, the most relevant subset is: $relevant")
```

Comment:
This could potentially generate pretty long log messages in bad clusters... perhaps just log the count?